tempesta_fw/ss_skb.c

```diff
@@ -527,70 +575,102 @@ static int
 __skb_fragment(struct sk_buff *skb, struct sk_buff *pskb, char *pspt,
 	       int len, TfwStr *it)
 {
-	int i, dlen;
-	char *vaddr;
-	struct sk_buff *frag_i;
-
-	BUG_ON(!len);
-	/* We can't modify data of shared or cloned skb. */
-	BUG_ON(skb_shared(skb) || skb_cloned(skb));
-
-	/* Determine where @split begins within socket buffer. */
-	dlen = skb_headlen(skb);
-	vaddr = skb->data;
+	unsigned int i, d_size, offset;
 
 	SS_DBG("skb fragmentation (len=%d pspt=%p, skb: head=%p data=%p"
 	       " tail=%p end=%p len=%u data_len=%u truesize=%u"
-	       " nr_frags=%u frag_list=%p)...\n",
+	       " nr_frags=%u)...\n",
 	       len, pspt, skb->head, skb->data,
 	       skb_tail_pointer(skb), skb_end_pointer(skb),
 	       skb->len, skb->data_len, skb->truesize,
-	       skb_shinfo(skb)->nr_frags, skb_shinfo(skb)->frag_list);
+	       skb_shinfo(skb)->nr_frags);
+	BUG_ON(!len);
+
+	if ((len > (int)PAGE_SIZE) || (-len > (int)PAGE_SIZE)) {
+		SS_WARN("SKB room to add or delete is too big: %u\n", len);
+		return -EINVAL;
+	}
 
-	if (pspt >= vaddr && pspt < vaddr + dlen) {
+	/*
+	 * Determine where the split begins within the SKB,
+	 * then do the job using the right function.
+	 */
+
+	/* See if the split begins in the linear data. */
+	d_size = skb_headlen(skb);
+	offset = pspt - (char *)skb->data;
+
+	if ((offset >= 0) && (offset < d_size)) {
 		it->ptr = __split_linear_data(skb, pskb, pspt, len);
 		goto done;
 	}
 
+	/* See if the split begins in the page fragments data. */
 	for (i = 0; i < skb_shinfo(skb)->nr_frags; ++i) {
 		const skb_frag_t *frag = &skb_shinfo(skb)->frags[i];
-		vaddr = skb_frag_address(frag);
-		dlen = skb_frag_size(frag);
+		d_size = skb_frag_size(frag);
+		offset = pspt - (char *)skb_frag_address(frag);
 
-		if (pspt >= vaddr && pspt < vaddr + dlen) {
-			it->ptr = __split_pgfrag(skb, pskb, i, pspt - vaddr,
-						 len);
+		if ((offset >= 0) && (offset < d_size)) {
+			it->ptr = __split_pgfrag(skb, pskb, i, offset, len);
 			goto done;
 		}
 	}
 
-	skb_walk_frags(skb, frag_i) {
-		int r = __skb_fragment(frag_i, skb, pspt, len, it);
-		if (r != -ENOENT)
-			return r;
-	}
```

Comment:
Why do you delete the code? ```ss_skb_defraglist()``` actually doesn't move the whole fragments list to ```TFW_SKB_CB()``` list. If you'd move the list, then you need to convert skb fragment list to Tempesta's list, which use different skb fields to organize the lists. This is double list traversals (firstly to convert the list and secondly to process the list) and there is no sense for the overhead, whereas the lists are typically short.